### PR TITLE
New resume method

### LINKED ISSUE #2739: Components should enter 'pause' filename as resume entry
Currently, the data ESP component (and future active ESP components) must get their information on a component's pause file(s) from that component's rpointer file. However, the data assimilation 'pause' file currently under design by CAM will not be a restart file but a special file containing only state fields needed by ESP components such as DART.
Components need an easy way to pass the name of the file written in response a pause signal to the ESP component.
At the same time, the ESP component needs to be able to have access to and gather these filenames from all component instances.

### PR DESCRIPTION
**Pull request by gold2718**
Re-implementation of pause/resume with a few features:

- Components specify a pause filename that does not have to be the same as the restart filename
- Components use of resume is now multi-driver friendly (no array of resume filenames)
- The post-assimilation check at initialization is now a simple logical

@ekluzek, @brian-eaton:  Note that this is an incompatible change for those components (CAM, CLM) that currently implement the post-assimilation check at initialization. The new check is included in a call to `seq_cdata_setptrs`:
```
call seq_cdata_setptrs(cdata, post_assimilation=foo)
```

After the call, `foo` is `.true.` if this restart is a post-data-assimilation run.

Test suite: scripts_regression_tests.py
Test baseline: NA
Test namelist changes: NA
Test status: bit for bit

Fixes #2739 

User interface changes?: No

Update gh-pages html (Y/N)?: N

Code review:

### CODE REVIEW

File: src/drivers/mct/cime_config/testdefs/testlist_drv.xml [hunk 1]
```diff
@@ -521,14 +521,6 @@
...
     </machines>
   </test>
-  <test name="PRE" grid="f19_f19" compset="ADESP_TEST">
-    <machines>
-      <machine name="hobart" compiler="nag" category="prealpha"/>
-    </machines>
-    <options>
-      <option name="wallclock"> 00:20 </option>
-    </options>
-  </test>
```
Comment (gold2718) [inline src/drivers/mct/cime_config/testdefs/testlist_drv.xml]: This test was duplicated below (see starting at line 627 of the new version).

File: src/drivers/mct/main/cime_comp_mod.F90 [hunk 2]
```diff
@@ -635,8 +637,6 @@ subroutine cime_pre_init1()
          iam=comp_comm_iam(it))
     if (iamroot_CPLID) output_perf = .true.
 
-    if (iamin_CPLID) complist = trim(complist)//' cpl'
-
```
Comment (gold2718) [inline src/drivers/mct/main/cime_comp_mod.F90]: This version of complist is not properly initialized and was never used (see initialization in cime_init starting at line 1214).

File: src/drivers/mct/main/component_mod.F90 [hunk 3]
```diff
@@ -137,61 +135,40 @@ subroutine component_init_pre(comp, compid, cplcompid, cplallcompid, &
...
-
-       ! Does this component need to do post-data assimilation processing?
-       if (seq_timemgr_data_assimilation_active(ntype(1:3))) then
-          comp_resume(:) = 'TRUE'
-       else
-          comp_resume(:) = ''
-       end if
+       call seq_cdata_init(comp(eci)%cdata_cc, comp(eci)%compid,              &
+            'cdata_'//ntype(1:1)//ntype(1:1), comp(eci)%dom_cc,               &
+            comp(eci)%gsmap_cc, infodata, seq_timemgr_data_assimilation_active(ntype(1:3)))
```
Comment (gold2718) [inline src/drivers/mct/main/component_mod.F90]: For some reason seq_cdata_init was not being used. I have resurrected it and added the new arguments.

File: src/drivers/mct/main/component_type_mod.F90 [hunk 4]
```diff
@@ -92,7 +92,6 @@ module component_type_mod
      logical                         :: iamroot_compid
      logical                         :: present ! true => component is present and not stub
      integer                         :: nthreads_compid
-     integer                         :: instn
      character(len=CL)               :: suffix
```
Comment (gold2718) [inline src/drivers/mct/main/component_type_mod.F90]: `instn` was never used but it looked useful (i.e., I tried to use it).

**Comment by jedwards4b** [comment]
@gold2718  can you offer branch tags for clm and cam which are compatible with this change?  Should I hold off merge until this can be coordinated?

**Comment by jedwards4b** [comment]
@gold2718  You would create branches in the clm and cam repos and put them and this update on the cesm2.2.x plans page.  indicating that there is a depandancy between them.

**Comment by ekluzek** [comment]
I've created a branch for this on my fork of ctsm. But, I'm running into trouble, so don't have it working yet. I'm pointing to @gold2718 cime branch for it right now.

**Pull request merged**